### PR TITLE
Remove `librustc_ast` session globals

### PR DESCRIPTION
**Pull request by nnethercote**
By moving the data onto `Session`.

r? @petrochenkov

### CODE REVIEW

**Comment by bjorn3** [comment]
This is a slight (up to 0.4%) slowdown.

File: src/librustc_metadata/creader.rs [hunk 1]
```diff
@@ -935,3 +936,26 @@ impl<'a> CrateLoader<'a> {
         self.maybe_resolve_crate(name, DepKind::Explicit, None).ok()
     }
 }
+
+pub fn global_allocator_spans(sess: &Session, krate: &ast::Crate) -> Vec<Span> {
```
Comment (petrochenkov) [inline src/librustc_metadata/creader.rs:940]:
```suggestion
fn global_allocator_spans(sess: &Session, krate: &ast::Crate) -> Vec<Span> {
```

File: src/librustc_hir/Cargo.toml [hunk 2]
```diff
@@ -16,6 +16,7 @@ rustc_data_structures = { path = "../librustc_data_structures" }
 rustc_index = { path = "../librustc_index" }
 rustc_span = { path = "../librustc_span" }
 rustc_serialize = { path = "../librustc_serialize" }
+rustc_session = { path = "../librustc_session" }
```
Comment (petrochenkov) [inline src/librustc_hir/Cargo.toml:19]: Removing this dependency was a conscious decision, IIRC.

**Comment by petrochenkov** [comment]
I suspect that majority of these `check_names` don't actually need to mark the attributes as used.
So were not only doing extra work, but now also need to pass `Session` to locations where it shouldn't otherwise be.
I need to check a couple of things.

**Comment by petrochenkov** [comment]
>I need to check a couple of things.

Blocked on https://github.com/rust-lang/rust/pull/74952.

**Comment by nnethercote** [comment]
> I suspect that majority of these `check_names` don't actually need to mark the attributes as used.

I was wondering about that, but I had no sense of which were necessary and which weren't, so I kept them all except for the couple that I couldn't keep. It'd be great if there was a clear rule about when `check_name` is needed.

File: src/librustc_hir/Cargo.toml (hunk 2 above)
Comment (nnethercote) [reply src/librustc_hir/Cargo.toml:19]: Is there any documentation on which crates should/can depend on others?

There are only two functions in `librustc_hir` that call `check_name`, so avoiding this dependency hopefully won't be too hard.

**Comment by petrochenkov** [comment]
>I was wondering about that, but I had no sense of which were necessary and which weren't, so I kept them all except for the couple that I couldn't keep. It'd be great if there was a clear rule about when `check_name` is needed.

I started working on this and will make a PR tomorrow, most likely.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/74952 ended up requiring too much manual work and included some nontrivial cases, so I submitted a simpler https://github.com/rust-lang/rust/pull/75043 instead.

File: src/librustc_hir/Cargo.toml (hunk 2 above)
Comment (petrochenkov) [reply src/librustc_hir/Cargo.toml:19]: No documentations, https://github.com/rust-lang/rust/pull/68133 just broke dependencies that were easily breakable.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/75043 has landed.

**Comment by nnethercote** [comment]
I have updated. Most of the changes outside of rustc are now no longer necessary, but the changes within rustc are still there. So it's still a big change.

File: src/librustc_ast/attr/mod.rs [hunk 3]
```diff
@@ -10,72 +10,30 @@ use crate::ptr::P;
...
-            // We have no idea how many attributes there will be, so just
-            // initiate the vectors with 0 bits. We'll grow them as necessary.
-            used_attrs: Lock::new(GrowableBitSet::new_empty()),
-            known_attrs: Lock::new(GrowableBitSet::new_empty()),
-            span_session_globals: rustc_span::SessionGlobals::new(edition),
-        }
+impl MarkedAttrs {
+    // We have no idea how many attributes there will be, so just
+    // initiate the vectors with 0 bits. We'll grow them as necessary.
+    pub fn new() -> Self {
```
Comment (petrochenkov) [inline src/librustc_ast/attr/mod.rs:26]: You could derive `Default` instead.

File: src/librustc_ast_passes/feature_gate.rs [hunk 4]
```diff
@@ -1,40 +1,44 @@
...
 use rustc_span::symbol::{sym, Symbol};
 use rustc_span::Span;
 
 use tracing::debug;
 
 macro_rules! gate_feature_fn {
-    ($cx: expr, $has_feature: expr, $span: expr, $name: expr, $explain: expr) => {{
-        let (cx, has_feature, span, name, explain) = (&*$cx, $has_feature, $span, $name, $explain);
-        let has_feature: bool = has_feature(&$cx.features);
+    ($vis: expr, $has_feature: expr, $span: expr, $name: expr, $explain: expr) => {{
```
Comment (petrochenkov) [inline src/librustc_ast_passes/feature_gate.rs:16]: Could you keep `cx`? (Or use `visitor` or `v`.)
`vis` is usually used for visibilities, so this looks confusing.

File: src/librustc_session/session.rs [hunk 5]
```diff
@@ -1283,6 +1358,8 @@ pub fn build_session(
         real_rust_source_base_dir,
         asm_arch,
         target_features: FxHashSet::default(),
+        known_attrs: Lock::new(MarkedAttrs::new()),
+        used_attrs: Lock::new(MarkedAttrs::new()),
```
Comment (petrochenkov) [inline src/librustc_session/session.rs:1362]: `Default::default()` can be used in both cases.

File: src/librustc_span/lib.rs [hunk 6]
```diff
@@ -87,6 +87,15 @@ impl SessionGlobals {
     }
 }
 
+pub fn with_session_globals<R>(edition: Edition, f: impl FnOnce() -> R) -> R {
+    let session_globals = SessionGlobals::new(edition);
+    SESSION_GLOBALS.set(&session_globals, f)
+}
+
+pub fn with_default_session_globals<R>(f: impl FnOnce() -> R) -> R {
+    with_session_globals(crate::edition::DEFAULT_EDITION, f)
```
Comment (petrochenkov) [inline src/librustc_span/lib.rs:96]:
```suggestion
    with_session_globals(edition::DEFAULT_EDITION, f)
```

File: src/librustc_hir/Cargo.toml (hunk 2 above)
Comment (petrochenkov) [reply src/librustc_hir/Cargo.toml:19]: I still think it would be good to break this dependency.

File: src/librustc_ast/attr/mod.rs (hunk 3 above)
Comment (nnethercote) [reply src/librustc_ast/attr/mod.rs:26]: `GrowableBitSet` doesn't derive `Default`; it has a `new_empty()` method. And it's built on top of `BitSet`, which also doesn't derive `Default`, and has both `new_empty()` and `new_filled()` methods.

File: src/librustc_ast_passes/feature_gate.rs (hunk 4 above)
Comment (nnethercote) [reply src/librustc_ast_passes/feature_gate.rs:16]: I will use `$visitor`.

File: src/librustc_hir/Cargo.toml (hunk 2 above)
Comment (nnethercote) [reply src/librustc_hir/Cargo.toml]: I have removed the dependency by passing in closures that capture the `Session` to two functions.

**Comment by nnethercote** [comment]
I have addressed all the comments and updated.

File: src/librustc_ast/attr/mod.rs (hunk 3 above)
Comment (petrochenkov) [reply src/librustc_ast/attr/mod.rs:26]: Ah, sorry, confused it with `FiniteBitSet` (which does implement `Default`).

**Comment by nnethercote** [comment]
It's a weird rustdoc failure I haven't seen before, only on one of the linux platforms:
```
error[E0275]: overflow evaluating the requirement `alloc::raw_vec::RawVec<(rustc_ast::ast::UseTree, rustc_ast::ast::NodeId)>: std::marker::Sync`
  |
  = help: consider adding a `#![recursion_limit="256"]` attribute to your crate (`rustc_plugin_impl`)
  = note: required because it appears within the type `std::vec::Vec<(rustc_ast::ast::UseTree, rustc_ast::ast::NodeId)>`
  = note: required because it appears within the type `rustc_ast::ast::UseTreeKind`
  ... <lots more lines like that> ...
```
This is while running rustdoc on the `rustc_plugin_impl` module. That seems unrelated to my changes, unless we happened to be really close to the recursion limit previously and my patch changed enough things to tip it over the limit?

I'm not sure what to do here, whether it's spurious and likely to disappear if we try again, or if it's likely to recur. If it's the latter, I guess I should add the `recusion_limit="256"` like the message suggests?

@jyn514, any ideas?

**Comment by jyn514** [comment]
I wouldn't expect it to be random. I'm not sure how it's related to your change, but it doesn't seem to be an infinite loop so I would try raising the type limit like it suggests.

**Comment by nnethercote** [comment]
I added the 256 recursion limit.

@bors r=petrochenkov

**Pull request merged**